REF/ENH: Refactor NDFrame finalization

TomAugspurger opened the pull request:
In preperation for https://github.com/pandas-dev/pandas/issues/27108
(disallowing duplicates), we need to enhance our metadata propagation.

*We need a way for a particiular attribute to deterimine how it's
propagated for a particular method*. Our current method of metadata
propagation lacked two features

1. It only copies an attribute from a source NDFrame to a new NDFrame.
   There is no way to propagate metadata from a collection of NDFrames
   (say from `pd.concat`) to a new NDFrame.
2. It only and always copies the attribute. This is not always
   appropriate when dealing with a collection of input NDFrames, as the
   source attributes may differ. The resolution of conflicts will differ
   by attribute (for `Series.name` we might throw away the name. For
   `Series.allow_duplicates`, any Series disallowing duplicates should
   mean the output disallows duplicates)

- [ ] Closes https://github.com/pandas-dev/pandas/issues/8572
- [ ] Update existing calls to `__finalize__ ` to pass through `method`
- [ ] profile / Run ASV

TomAugspurger committed:
REF/ENH: Refactor NDFrame finalization

In preperation for https://github.com/pandas-dev/pandas/issues/27108
(disallowing duplicates), we need to enhance our metadata propagation.

*We need a way for a particiular attribute to deterimine how it's
propagated for a particular method*. Our current method of metadata
propagation lacked two features

1. It only copies an attribute from a source NDFrame to a new NDFrame.
   There is no way to propagate metadata from a collection of NDFrames
   (say from `pd.concat`) to a new NDFrame.
2. It only and always copies the attribute. This is not always
   appropriate when dealing with a collection of input NDFrames, as the
   source attributes may differ. The resolution of conflicts will differ
   by attribute (for `Series.name` we might throw away the name. For
   `Series.allow_duplicates`, any Series disallowing duplicates should
   mean the output disallows duplicates)

TomAugspurger commented:
cc @jreback @jbrockmendel, this is the start of what I need for https://github.com/pandas-dev/pandas/issues/27108#issuecomment-527633246. Briefly, adding a `.allows_duplicate_labels` attribute to NDFrame, which will be propagated through all operations. That already works for most things, but it's dropped for operations involving multiple inputs (concat, binops, etc.).

This PR puts the infrastructure in place to support this. We aren't using it yet (outside of tests), but will when I make the duplicate labels PR.

```python
duplicate_labels_meta = PandasMetadata("allows_duplicates")


@duplicate_labels_meta.register(pd.concat)
def _(new, concatenater):
    new.allows_duplicate_labels = all(x.allows_duplicate_labels for x in concatenater.objs)

@duplicate_labels_meta.register(Series.__array_ufunc__)
def _(new, inputs):
    new.allows_duplicate_labels = ...
```

But I split this off so that PR will be smaller.

WillAyd commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+from pandas.core.dtypes.generic import ABCDataFrame, ABCSeries
+
+if TYPE_CHECKING:
+    from pandas.core.generic import NDFrame
```
I *think* you want FrameOrSeries from pandas._typing

WillAyd commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+# Pandas Internals.
+
+
+def ndframe_finalize(new: "NDFrame", other: Any, method: dispatch_method_type):
```
Going back to previous comment, if you use `FrameOrSeries` here it is a TypeVar, so you could type it like `def ndframe_finalize(new: FrameOrSeries, ...) -> FrameOrSeries` which would guarantee the type of new and the type of the return value are the same (i.e. function becomes generic)

WillAyd commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+    Parameters
+    ----------
+    new : NDFrame
+    other : NDFrame
```
I guess this conflicts with the annotation of `Any` - can you speak to what other types you think would be acceptable here?

TomAugspurger replied:
Sorry it's supposed to be Any.

Methods like `pd.concat` pass a `_Concatenator` here. I think `Series.__array_ufunc__` will pass the tuple `inputs`.

WillAyd commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+if TYPE_CHECKING:
+    from pandas.core.generic import NDFrame
+
+dispatch = defaultdict(dict)
```
Can you annotate this? Would help read what should be nested therein

TomAugspurger replied:
I *think* it's supposed to be `DefaultDict[dispatch_method_type, Callable[..., FrameOrSeries]]`

jbrockmendel commented on pandas/tests/generic/test_metadata.py:
```diff
@@ -0,0 +1,49 @@
+    new.attr = other.attr + 1
+
+
+@mymeta.register("concat")
```
an alternative to this registering which might be more self-contained is subclassing PandasMetadata and implementing/overriding the relevant methods/functions

TomAugspurger replied:
Hmm, yes... I'll think about this. This may fix the note I had in the module docstring that we don't document pandas calls actually call `__finalize__`, and what they're called with.

jbrockmendel commented:
Not necessarily a precursor to this, but it would be nice if we had a systematic way to make sure we were using `_constructor` and `__finalize__` in all the appropriate places.

TomAugspurger commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+
+        Parameters
+        ----------
+        pandas_method : callable or str
```
The idea was to register one finalizer per pandas method. I Brock's subclassed based approach will make this clearer.

Pandas will provide a default implementation, which the subclass can override.

TomAugspurger commented on pandas/core/_meta.py:
```diff
@@ -0,0 +1,146 @@
+        finalizer = dispatch.get(key_of(method), {}).get(name)
+
+        if finalizer:
+            finalizer(new, other)
```
Just a style choice. All these operations are inplace. My hope is that by returning `None`, we make it clearer that you can't return a new object.

TomAugspurger committed:
refactor

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into metadata-dispatch

TomAugspurger commented:
> it would be nice if we had a systematic way to make sure we were using _constructor and __finalize__ in all the appropriate places.

I refactored to use a class-based approach, rather than decorators. The weirdest thing is that users need to actually create an instance of their subclass so that we know about it, but don't actually do anything with it.

```python
>>> mymeta = MyMeta("mymeta")
```

Now that I think about it, I guess that we could have a class decorator for registering. So then they do something like

```python

@register_metadata
class MyMeta(PandasMetadata):
    name = "mymeta"
```

And then we just instantiate it when we need it, rather than relying on singletons per `name`.

TomAugspurger committed:
move to meta

jbrockmendel commented:
> The weirdest thing is that users need to actually create an instance of their subclass so that we know about it, but don't actually do anything with it.

Yah, this is a little confusing.  What if my Meta subclass can take on multiple values, I'd assume I'd need multiple instances of it.

TomAugspurger commented:
> What if my Meta subclass can take on multiple values

Do you mean refer to multiple attributes? If so, then yes, that would get strange.


```python
class MyMetaBase(PandasMetadata):
    ...

@register_metdata
class MyMyetaA(MyMetaBase):
    name = 'a'

@register_metadata
class MyMetaB(MyMetaBase):
    name = 'b
```

versus

```python
my_meta_a = MyMetaBase(name='a')
my_meta_b = MyMetaBase(name='b')
```

neither is great.

TomAugspurger committed:
mypy

jbrockmendel commented:
> Do you mean refer to multiple attributes? If so, then yes, that would get strange.

I mean different values that a single attribute can take.  e.g.

```
class AllowsDuplicateLabels(PandasMetadata):
    def __init__(self, allows):
        self.allows = allows

    def __bool__(self):
        return self.allows

    [...]


ser = pd.Series(range(3)
ser._metadata["allows"] = AllowsDuplicateLabels(False)

ser2 = pd.Series(range(4))
ser2._metadata["allows"] = AllowsDuplicateLabels(True)

>>> ser.index = [1, 1, 1]
ValueError: Series does not allow duplicate labels.
```

TomAugspurger commented:
Hmm OK. Not sure I fully understand, but let me try to clarify.

Everything I've put up here as to do with metadata *resolution*: How do we transfer metadata from the source object(s) to the new object. It looks like your example is getting into specifying metadata values, though I may be misunderstanding. I haven't thought much about changing `NDFrame._metadata` from a list of strings specifying attributes.

jbrockmendel commented:
> I haven't thought much about changing NDFrame._metadata from a list of strings specifying attributes.

That part of my example was probably unhelpful.  If it makes it any clearer, pretend I use something other than `_metadata` that doesn't already exist.

TomAugspurger committed:
fixed subclass

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into metadata-dispatch

TomAugspurger commented:
Sorry, I'm still not following your example.

TomAugspurger committed:
define API

TomAugspurger committed:
simplify

TomAugspurger commented:
Changed my mind one more time. I started down a class based approach where we define

```
class PandasMeta:
    def copy(self, new, other):
    def concat(self, new, other):
    def sort_index(self, new, other):
```

While that's nice, since it gives a well-defined API for what finalize does, it complicates the implementation of the common case a decent amount. You can see it at https://github.com/pandas-dev/pandas/compare/master...TomAugspurger:metadata-dispatch+complex?expand=1

So I've backed off that and simplified this to get things working for my short-term needs (disallow_duplicates). A piece of metadata still has per-method control (pandas will need to be updated to pass `method=` in more places).

Doing some perf checks now.

jorisvandenbossche commented:
Do we need this class-based approach? What can you not do with the current implementation?

> It only copies an attribute from a source NDFrame to a new NDFrame.
    There is no way to propagate metadata from a collection of NDFrames
    (say from pd.concat) to a new NDFrame.

This is already possible, as `__finalize__` gets passed the `_concatenator` object, which has access to all frames passed to concat. 
This is what we do in geopandas (https://github.com/geopandas/geopandas/blob/29add0a735b00dc20c79e0fccc8e6a775c4997b0/geopandas/geodataframe.py#L561-L574)

>It only and always copies the attribute. This is not always
    appropriate when dealing with a collection of input NDFrames, as the
    source attributes may differ. The resolution of conflicts will differ
    by attribute

In principle, you can do this in `__finalize__` as well. You have access to the name of the attribute, so you can implement different logic for different attributes. It might not necessarily nice to have to put this all in `__finalize__`, and another approach could be cleaner. But I don't fully understand what *functionality* you are trying to add.

jorisvandenbossche commented:
To be clear, I am not opposed to making the implementation nicer, or making it more easily extensible to add custom logic, etc. I just first want to understand the constraints / needed functionality, before forming my opinion on the new implementation.

An easier way for external people to define custom behaviour for metadata without clashing with pandas (eg currently by overriding `__finalize__`, if pandas would add custom logic for an `allow_duplicates`, the current implementation of geopandas would override that I think), would certainly be very valuable!

jbrockmendel commented on pandas/tests/generic/test_metadata.py:
```diff
@@ -0,0 +1,58 @@
+from pandas.core.meta import PandasMetadata
+
+
+class MyMeta(PandasMetadata):
```
Do you have an implementation for index_allows_duplicates?  I'd be more comfortable if there were a more fully fleshed-out example/test.

TomAugspurger replied:
Hmm, I thought I did but am having trouble finding it right now.

TomAugspurger commented:
@jorisvandenbossche I put up https://github.com/pandas-dev/pandas/pull/28394 for ease of comparison. That'll be the "minimal" approach (no general changes to `NDArray.__finalize__`)

TomAugspurger commented:
Right now, I'm going to push forward on https://github.com/pandas-dev/pandas/pull/28394. That may give some guidance on how to design a metadata finalization API. Apologies for the wasted review time.
